src/System.Linq/tests/DistinctTests.cs

```diff
@@ -164,5 +136,57 @@ public void CustomEqualityComparer()
 
             Assert.Equal(expected, source.Distinct(new AnagramEqualityComparer()), new AnagramEqualityComparer());
         }
+
+        [Theory, MemberData("SequencesWithDuplicates")]
+        public void FindDistinctAndValidate<T>(T unusedArgumentToForceTypeInference, IEnumerable<T> original)
+        {
+            // Convert to list to avoid repeated enumerations of the enumerables.
+            var originalList = original.ToList();
+            var distinctList = originalList.Distinct().ToList();
+
+            // Ensure the result doesn't contain duplicates.
+            var hashSet = new HashSet<T>();
+            foreach (var i in distinctList)
+            {
+                Assert.False(hashSet.Contains(i), "Found a duplicate in the result of Distinct(): " + i);
+                hashSet.Add(i);
+            }
+
+            // Ensure that every element in 'originalList' exists in 'distinctList'.
+            Assert.DoesNotContain(originalList, x => !distinctList.Contains(x));
+
+            // Ensure that every element in 'distinctList' exists in 'originalList'.
+            Assert.DoesNotContain(distinctList, x => !originalList.Contains(x));
```

Comment:
Good idea. On a related note, there's a lot of places that should be Assert.Empty but aren't.